src/python/pants/testutil/engine/util.py

```diff
@@ -143,6 +145,52 @@ def configurations(self):
 TARGET_TABLE = SymbolTable({'struct': Struct, 'target': Target})
 
 
+def rootify_rules(
+  *rules: Union[Rule, UnionRule]
+) -> Tuple[Union[TaskRule, RootRule, UnionRule], ...]:
+  """Utility to convert `optionable_rule`s into `RootRule`s and no-op on other types of rules.
+
+  This is useful for `TestBase.request_single_product_request`-style tests during the setup of
+  `rules()`. To use an `optionable_rule` in tests, you have to provide an `OptionsBootstrapper`,
+  which is more heavy-weight than most tests want. This prevents certain tests from using the
+  following style, because it means the test will now require `OptionsBootstrapper`:
+
+  ```
+  from pants.rules.core.fmt import rules as fmt_rules
+
+  class TestFmt(TestBase):
+
+    @classmethod
+    def rules():
+      return (*super.rules(), *fmt_rules())
+  ```
+
+  Instead, every `optionable_rule` would need to be registered as a `RootRule`. But, it is tedious
+  to have to register every individual Subsystem used transitively as a RootRule and to also
+  register every individual rule used transitively.
+
+  Instead, this utility function allows for this:
+
+  ```
+  from pants.rules.core.fmt import rules as fmt_rules
+  from pants.testutil.engine.util import rootify_rules
+
+  class TestFmt(TestBase):
+
+    @classmethod
+    def rules():
+      return rootify_rules(*super.rules(), *fmt_rules())
+  ```
+  """
+  def is_optionable_rule(rule: Union[Rule, UnionRule]) -> bool:
+    return hasattr(rule, "output_type") and issubclass(rule.output_type, Subsystem)
```

Comment:
How about `isinstance(rule, Rule)` or `isinstance(rule, TaskRule)`. It seems a bit strange to dip into reflection when you have types.

Reply:
True. 

The current type hints lie in saying that `rootify_rules` takes `Union[Rule, UnionRule]`. It also takes `Callable`, which is the type for rule functions like `fmt` and `inject_init`. I'll fix those too.